### PR TITLE
Highlight assembly line on source selected line change

### PR DESCRIPTION
**Pull request by hatal175**
This could really use a good code review from someone who knows react.

### CODE REVIEW

File: frontend/src/components/Editor/CodeMirror.tsx [hunk 1]
```diff
@@ -78,5 +85,25 @@ export default function CodeMirror({ value, onChange, className, viewRef: viewRe
         }
     }, [value])
 
-    return <div ref={el} className={className} style={{ fontSize: "0.8em" }} />
+    const debounceHover = debounce(50, false, (event: MouseEvent) => {
```
Comment (nanaian) [inline frontend/src/components/Editor/CodeMirror.tsx:88]:
```suggestion
    const debouncedOnMouseMove = debounce(50, false, (event: MouseEvent) => {
```

File: frontend/src/components/Editor/CodeMirror.tsx [hunk 2]
```diff
@@ -27,6 +29,11 @@ export default function CodeMirror({ value, onChange, className, viewRef: viewRe
     const extensionsRef = useRef(extensions)
     extensionsRef.current = extensions
 
+    const hoveredSourceLineref = useRef<number>()
```
Comment (nanaian) [inline frontend/src/components/Editor/CodeMirror.tsx:32]:
```suggestion
    const hoveredSourceLineRef = useRef<number>()
```

File: frontend/src/components/Editor/CodeMirror.tsx [hunk 3]
```diff
@@ -2,18 +2,20 @@ import { MutableRefObject, useEffect, useRef } from "react"
...
 import { Extension, EditorState } from "@codemirror/state"
 import { EditorView } from "@codemirror/view"
+import { debounce } from "throttle-debounce"
 
 import { materialPalenight } from "../../lib/themes/dark"
 
 export interface Props {
     value: string
     onChange?: (value: string) => void
+    OnHoverSourceChange?: (value: number | null) => void
```
Comment (nanaian) [inline frontend/src/components/Editor/CodeMirror.tsx:12]:
```suggestion
    onHoveredSourceLineChange?: (value: number | null) => void
```

File: frontend/src/components/Diff/Diff.tsx [hunk 4]
```diff
@@ -24,19 +24,20 @@ function FormatDiffText({ texts }: { texts: api.DiffText[] }) {
...
+    hoveredSourceLine?: number | null
 }) {
     return <resizer.Section className={classNames(styles.column, className)} minSize={100}>
         <div className={classNames(styles.row, styles.header)}>
             {header}
         </div>
         <div className={styles.body}>
             {diff?.rows?.map?.((row, i) => (
-                <div key={i} className={styles.row}>
+                <div key={i} className={(typeof row[prop]?.src_line != "undefined" && row[prop]?.src_line == hoveredSourceLine) ? styles.hoveredSourceRow : styles.row}>
```
Comment (nanaian) [inline frontend/src/components/Diff/Diff.tsx:40]:
```suggestion
                <div
                    key={i}
                    className={classNames({
                        [styles.row]: true,
                        [styles.highlight]: typeof hoveredSourceLine == "number" && row[prop]?.src_line == hoveredSourceLine}
                    })}
                >
```

File: frontend/src/components/Diff/Diff.module.scss [hunk 5]
```diff
@@ -29,6 +29,12 @@
     height: 1.25em;
 }
 
+.hoveredSourceRow {
+    padding: 0 4px;
+    height: 1.25em;
+    background-color: var(--a150);
```
Comment (nanaian) [inline frontend/src/components/Diff/Diff.module.scss:35]:
```suggestion

    &.highlight {
        background-color: var(--a150);
    }
```

File: frontend/src/lib/api.ts [hunk 6]
```diff
@@ -192,6 +192,7 @@ export interface Scratch extends TerseScratch {
     context: string
     diff_label: string
     parent: string | null
+    hoveredSourceLine: number | null
```
Comment (nanaian) [inline frontend/src/lib/api.ts:195]: This piece of state should not be contained in Scratch.

**Review by nanaian — CHANGES_REQUESTED** [review]
Nice PR! There are some issues with where state lives right now...

`hoveredSourceLine` state should not be contained in `Scratch`, for a couple reasons:
1. `Scratch` is directly reflective of what the API returns and expects
2. `hoveredSourceLine` is _editor state_ and not relevant to the rest of the scratch data (besides `source_code`)

Instead, you should do `const [hoveredSourceLine, setHoveredSourceLine] = useState<number | null>(null)` in the `<Scratch>` component (`frontend/src/components/Scratch/Scratch.tsx`) and pass down`setHoveredSourceLine` and/or `hoveredSourceLine` where required. For instance, CompilationPanel can just take `hoveredSourceLine` instead of the whole scratch, and it shouldn't need its own copy of that state or that useEffect handler.

If you need any help making these changes feel free to drop me a message on Discord :shipit:

File: frontend/src/components/Diff/Diff.tsx (hunk 4 above)
Comment (nanaian) [reply frontend/src/components/Diff/Diff.tsx:40]: You may need to `import classNames from "classnames"`

Comment (hatal175) [reply frontend/src/components/Diff/Diff.tsx:40]: I haven't really changed anything here regarding classNames though...

**Review by nanaian — APPROVED** [review]

**Pull request merged**